### PR TITLE
Add job_schedulers and buildkite modules, move code out of `poll.py`

### PR DESCRIPTION
**Pull request by nefrathenrici**
Changes:
- `bin/buildkite.py`: Provides buildkite polling functions and environment variables, moved from `bin/poll.py`. I am not sure if there is a more pythonic way of passing these constants.
- `bin/job_schedulers.py`: Provides cluster job lifecycle control for the polling loop. Contains two classes `SlurmJobScheduler` and `PBSJobScheduler`, which each provide methods: `submit_job`, `cancel_jobs`, and `current_jobs`.
- `bin/poll.py`: Moved buildkite polling functions to `bin/buildkite.py`, moved slurm job submission to `bin/job_schedulers.py`.
- `bin/schedule_job.sh` New version of `bin/slurmjob.sh` but without any slurm dependencies.
`environment/hooks`: Add a test queue to debug or test new features without interrupting CI.

The new python files should probably be moved out of `bin`, since they are not executables.

### CODE REVIEW

File: bin/cron.sh [hunk 1]
```diff
@@ -10,12 +9,26 @@ case "$(hostname)" in
...
-    ;;
+        ;;
 esac
 
+case "$(hostname)" in
+    derecho[0-7]|"cron")
+        source /etc/bash.bashrc
+        ;;
+    *)
+        source /etc/bashrc
```
Comment (nefrathenrici) [inline bin/cron.sh:27]: add check if file exists instead

File: bin/cron.sh [hunk 2]
```diff
@@ -10,12 +9,30 @@ case "$(hostname)" in
         export BUILDKITE_PATH="/clima/slurm-buildkite"
         export BUILDKITE_QUEUE='clima'
         ;;
+    derecho[0-7]|"cron")
```
Comment (Sbozzolo) [inline bin/cron.sh:12]: `cron` could be a general name used by other computers as well, which might lead to name collision.

File: bin/job_schedulers.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+            with dbm.open(DATABASE_FILE, 'c') as db:
+                for k in db.keys():
+                    current_jobs[k.decode()] = db[k].decode()
+        except dbm.error as e:
+            logger.error(f"Failed to read from database: {e}")
+            return {}
+
+        try:
+            # The server is derecho-specific, could be abstractted out
+            qstat_output = subprocess.check_output(['qstat', '@desched1'], universal_newlines=True)
```
Comment (Sbozzolo) [inline bin/job_schedulers.py:230]: Yes, it would be good to abstract this out

File: bin/job_schedulers.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        try:
+            with dbm.open(DATABASE_FILE, 'w') as db:
+                for job_id in list(current_jobs.values()):
+                    if job_id not in active_pbs_jobs:
+                        logger.debug(f"Removing completed job from database: {job_id}")
+                        # TODO: Fix this awful loop...
+                        for k in db.keys():
+                            if db[k].decode() == job_id:
+                                del current_jobs[k.decode()]
+                                del db[k]
```
Comment (Sbozzolo) [inline bin/job_schedulers.py:253]: You can add a `break` so that this operation stops when the match is found

**Comment by nefrathenrici** [comment]
Todo: Update readme

File: bin/cron.sh (hunk 2 above)
Comment (nefrathenrici) [reply bin/cron.sh:12]: I added a check for the BUILDKITE_PATH, but I am not sure how to get identifying information on the derecho cron node.
```
if [[ -d BUILDKITE_PATH ]]; then
    echo "Could not find buildkite dir $BUILDKITE_PATH. Exiting..."
    exit 1
fi
```

Comment (Sbozzolo) [reply bin/cron.sh:12]: I don't know either. We can keep it as it, maybe just add a note to draw attention.

File: set_up_guide.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+Ensure that previously created TMPDIRs. As before, this can either be done with  `pdsh`, `srun` (with Slurm), or `pbsdsh` (with PBS).
+
+#### Optional: hooks/pre-command
+- Ensure that output folders have proper permissions
+
+### Testing
+
+You can test this setup by setting the `BUILDKITE_QUEUE` to `test` and manually running the cronjob `bin/cron.sh`. Logs from the cron job will be written to `logs/YYYY-MM-DD/cron`. 
+
+To test CUDA-aware MPI, you can run the testing script `bin/test_cuda_mpi.jl` within an MPI job. This currently tests for CUDA devices, MPI, working MPI with CUDA, and working MPI with CUDA ensuring peer-to-peer transfer.
```
Comment (Sbozzolo) [inline set_up_guide.md:67]: Let's update this in the future

**Review by Sbozzolo — APPROVED** [review]

**Pull request merged**